haskell/haskell-impl.bzl

```diff
@@ -0,0 +1,192 @@
+"""Implementation of core Haskell rules"""
+
+load(":actions.bzl",
+  "compile_haskell_bin",
+  "link_haskell_bin",
+  "compile_haskell_lib",
+  "link_static_lib",
+  "link_dynamic_lib",
+  "create_ghc_package",
+  "gather_dep_info",
+  "get_pkg_id",
+)
+load(":set.bzl",
+  "set",
+)
+load(":providers.bzl",
+  "HaskellBuildInfo",
+  "HaskellBinaryInfo",
+  "HaskellLibraryInfo",
+)
+load(":ghci-repl.bzl",
+  _build_haskell_repl = "build_haskell_repl",
+)
+
+def haskell_binary_impl(ctx):
+  c = compile_haskell_bin(ctx)
+
+  binary = link_haskell_bin(ctx, c.object_dyn_files)
+  dep_info = gather_dep_info(ctx)
+
+  solibs = set.union(
+    set.from_list(dep_info.external_libraries.values()),
+    dep_info.dynamic_libraries,
+  )
+
+  build_info = dep_info # HaskellBuildInfo
+  bin_info = HaskellBinaryInfo(
+    source_files = c.source_files,
+    modules = c.modules,
+    binary = binary,
+  )
+  target_files = depset([binary])
+
+  _build_haskell_repl(
+    ctx,
+    build_info=build_info,
+    target_files=target_files,
+    bin_info=bin_info,
+  )
+
+  return [
+    build_info,
+    bin_info,
+    DefaultInfo(
+      executable = binary,
+      files = target_files,
+      runfiles = ctx.runfiles(
+        files = set.to_list(solibs),
+        collect_data = True,
+      ),
+    ),
+  ]
+
+def haskell_library_impl(ctx):
```

Comment:
This approach still feels fragile; for regular libraries you're passing an actual `ctx`, but for proto libraries you're passing a struct with specific attributes.  At minimum we should document what attributes this function expects.

However, I additionally feel we should make two structural modifications: (1) separate this function from the rules's actual `haskell_library_impl` (as I mentioned in my comment here: https://github.com/tweag/rules_haskell/pull/239#discussion_r187375274), and (2) distinguish the "real" ctx from attributes that are being passed in manually.

For example:

```
def build_haskell_library(ctx, hs_files, attrs):
    # Document the fields expected in attrs,
    # which should only need to be things that are not already available from `ctx.attrs`
    # in both haskell_library and `haskell_proto_library`
    ...

def _haskell_library_impl(ctx):
    build_haskell_library(
        ctx,
       ctx.files.srcs,
        struct(...))
)

def _haskell_proto_library_impl(
    # ...
    hs_srcs = ...
    build_haskell_library(ctx, hs_srcs, struct(...))
)

haskell_library = rule(
    implementation=_haskell_library_impl,
    ...)
haskell_proto_library = rule(
    implementation=_haskell_proto_library_impl,
    ...)
```

Let me know what you think.

Reply:
I agree that this is fragile. I'll do the change you describe, however note that there is no way to forbid a developer accessing `deps` for example from context attributes, like so: `ctx.attr.deps` and not from that additional attribute that we'll be passing around. Test suite may not catch these things and they will lead to bugs. When I proposed the solution with patched `cxt` I kept in mind @mboes's comment that there should be only one way to access something. With the proposed approach I'll always be able to access things in two ways: from context `ctx` and from the additional attribute struct. I'm not even sure how to detect all the places where I need to do changes now.

Reply:
Yeah, those are fair points; I'm fine doing it either way for now.  (Who knew that dynamically-typed code could be hard to reason about...)